src/Microsoft.OData.Core/UriParser/Parsers/TypeParsers/UriCustomTypeParsers.cs

```diff
@@ -0,0 +1,243 @@
+                }
+
+                // Parse with all the general parsers
+                // Stop when a parser succeeded parsing the text.
+                foreach (IUriTypeParser uriCustomTypeParser in CustomUriTypeParsers)
+                {
+                    // Try to parse
+                    targetValue = uriCustomTypeParser.ParseUriStringToType(text, targetType, out parsingException);
+
+                    // The uriCustomParser could parse the given targetType but failed during the parsing proccess
+                    if (parsingException != null)
+                    {
+                        return null;
+                    }
+
+                    // In case of no exception and no value - The parse cannot parse the given text
+                    if (targetValue != null)
+                    {
+                        return targetValue;
+                    }
+                }
+            }
+
+            // No uriCustomParser could parse the requested uri text.
+            parsingException = null;
+            return null;
+        }
+
+        #endregion
+
+        #region Public Static Methods
+
+        /// <summary>
+        /// Add a custom 'IUriTypeParser' which will be called to parse uri values during the uri parsing proccess.
+        /// </summary>
+        /// <param name="uriCustomTypeParser">The custom uri parser</param>
+        /// <exception cref="ArgumentNullException">uriCustomTypeParser is null</exception>
+        /// <exception cref="ODataException">The given IUriTypeParser instance already exists</exception>
+        public static void AddCustomUriTypeParser(IUriTypeParser uriCustomTypeParser)
+        {
+            ExceptionUtils.CheckArgumentNotNull(uriCustomTypeParser, "uriCustomTypeParser");
+
+            lock (Locker)
+            {
+                if (CustomUriTypeParsers.Contains(uriCustomTypeParser))
+                {
+                    throw new ODataException(ODataErrorStrings.UriCustomTypeParsers_AddCustomUriTypeParserAlreadyExists);
+                }
+
+                CustomUriTypeParsers.Add(uriCustomTypeParser);
+            }
+        }
+
+        /// <summary>
+        /// Add a custom 'IUriTypeParser' which will be called to parse a value of the given EdmType during the UriParsing proccess.
+        /// </summary>
+        /// <param name="edmTypeReference">The EdmType the UriTypeParser can parse.</param>
+        /// <param name="uriCustomTypeParser">The custom uri type parser to add.</param>
+        /// <exception cref="ArgumentNullException">UriCustomTypeParser is null.</exception>
+        /// <exception cref="ArgumentNullException">EdmTypeReference is null.</exception>
+        /// <exception cref="ODataException">Another UriTypeParser is already registered for the given EdmType</exception>
+        public static void AddCustomUriTypeParser(IEdmTypeReference edmTypeReference, IUriTypeParser uriCustomTypeParser)

```

Comment:
I have implemented the API 'AddCustomUriTypeParser(IEdmTypeReference edmTypeReference, IUriTypeParser uriCustomTypeParser)' mainly for performance reasons.

I don't want taht in every parse, for each type we will call each custom parse. Moreover, I think that in most cases, the developers will register their CustomParser to one EdmType.
I know that I haven't implemented the registered custom parses with Dictionary(so it would be much more effective), and I have commented why.

In addition, I think the developers will appreciate it for being dynamic. One API for simplicity and the other one which let them handle everything by themselves.